### PR TITLE
ci(jenkins): enable opbeans

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

Enable the apm-agent PRs validation by triggering a build in the opbeans that evaluate if the consumer, aka the opbeans, behaves as expected. For such, it was required the below actions:
- Enable to checkout a sha commit.
- Add a couple of parallel stages to validate the opbeans and the opbeans-frontend.
- Create a couple of scripts to put the specific list of compose.py flags.
- Build the opbeans-frontend docker image for the given agent version on the fly.

## Why is it important?

Shift left to reduce failure. The current implementation does only detect issues during the release process, in fact, as a post-release operation. This will help to detect any kind of issue earlier.

This approach will delegate the opbeans generation as it is and will inject the agent on the fly with the apm-integration-testing library

## Related issues

https://github.com/elastic/apm-integration-testing/pull/591 could be a good candidate for the RUM opbeans

## Follow-ups

- Enhance the test coverage for the opbeans with the same test framework we have done in cypress for the k8s clusters.

## Tests

- opbeans-dotnet

```
scripts/compose.py start 8.0.0 --with-agent-dotnet   --with-opbeans-dotnet \
     --dotnet-agent-version 460b0078f3d56f677d194c5cb650293dac7ba2db \
     --opbeans-dotnet-agent-branch 460b0078f3d56f677d194c5cb650293dac7ba2db \
     --no-apm-server-dashboards \
     --no-apm-server-self-instrument  \
     --apm-server-agent-config-poll=1s \
     --no-xpack-secure
```

Build -> [pipeline](https://apm-ci.elastic.co/job/apm-integration-tests-selector-mbp/job/PR-763/18/)

- opbeans-go

```bash
$ scripts/compose.py start 8.0.0 --with-agent-go-net-http --with-opbeans-go \
      --go-agent-version d208b14ccff4e340cc6e14bfcda43850767f2a64 \
      --opbeans-go-agent-branch d208b14ccff4e340cc6e14bfcda43850767f2a64 \
      --no-apm-server-dashboards \
      --no-apm-server-self-instrument  \
      --apm-server-agent-config-poll=1s \
      --no-xpack-secure
```

Build -> [pipeline](https://apm-ci.elastic.co/job/apm-integration-tests-selector-mbp/view/change-requests/job/PR-763/20/)

- opbeans-java

```bash
$ scripts/compose.py start 8.0.0 --with-agent-java-spring --with-opbeans-java \
      --java-agent-version 7443263684dfb8031709bbfbc150915b8b340f9b \
      --opbeans-java-agent-branch 7443263684dfb8031709bbfbc150915b8b340f9b \
      --no-apm-server-dashboards \
      --no-apm-server-self-instrument  \
      --apm-server-agent-config-poll=1s \
      --no-xpack-secure

```

Build -> [pipeline](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-integration-tests-selector-mbp/detail/PR-763/16/pipeline)

- opbeans-node

```bash
$ scripts/compose.py start 8.0.0 --with-agent-nodejs-express --with-opbeans-node \
      --nodejs-agent-package 'elastic/apm-agent-nodejs#9adfd939449a24c7d46d260f0927edbd1bd45556' \
      --opbeans-node-agent-branch 9adfd939449a24c7d46d260f0927edbd1bd45556 \
      --no-apm-server-dashboards \
      --no-apm-server-self-instrument  \
      --apm-server-agent-config-poll=1s \
      --no-xpack-secure
```

Build -> [pipeline](https://apm-ci.elastic.co/job/apm-integration-tests-selector-mbp/job/PR-763/17/)

- opbeans-python

```bash
scripts/compose.py start 8.0.0 --with-agent-python-flask --with-opbeans-python \
     --python-agent-package git+https://github.com/elastic/apm-agent-python.git@6e82a8f63a10a669c6f4bbd556af394f208262ec \
     --opbeans-python-agent-branch 6e82a8f63a10a669c6f4bbd556af394f208262ec \
     --no-apm-server-dashboards \
     --no-apm-server-self-instrument  \
     --apm-server-agent-config-poll=1s \
     --no-xpack-secure
```

Build -> [pipeline](https://apm-ci.elastic.co/job/apm-integration-tests-selector-mbp/view/change-requests/job/PR-763/19/)

- opbeans-ruby

```bash
$ scripts/compose.py start 8.0.0 --with-agent-ruby-rails --with-opbeans-ruby \
      --ruby-agent-version df283cf465b1ce0af3a57db4fa184df0b5764992 \
      --ruby-agent-version-state commit \
      --ruby-agent-repo elastic/apm-agent-ruby \
      --opbeans-ruby-agent-branch df283cf465b1ce0af3a57db4fa184df0b5764992 \
      --no-apm-server-dashboards \
      --no-apm-server-self-instrument  \
      --apm-server-agent-config-poll=1s \
      --no-xpack-secure
```

Build -> [pipeline](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-integration-tests-selector-mbp/detail/PR-763/15/pipeline)

- Opbeans-frontend

```bash
$ scripts/compose.py start 8.0.0 --with-opbeans-go \
      --no-apm-server-dashboards \
      --no-apm-server-self-instrument  \
      --apm-server-agent-config-poll=1s \
      --no-xpack-secure

$ docker run -d --rm --name=opbeans-frontend
   --network="apm-integration-testing" \
   -e ELASTIC_APM_JS_BASE_SERVICE_NAME=opbeans-rum \
   -e ELASTIC_APM_JS_BASE_SERVER_URL="http://localtesting_8.0.0_apm-server:8200" \
   -e ELASTIC_OPBEANS_API_SERVER="http://localtesting_8.0.0_opbeans-go:3000" \
   -p 3000:3000 \
   docker.elastic.co/observability-ci/it_opbeans-frontend_nginx

$ curl -s -o /dev/null -w "%{http_code}" http://0.0.0.0:3000/dashboard
$ docker logs opbeans-frontend | grep '/dashboard'
$ docker stop opbeans-frontend || true
```

Build -> [pipeline](https://apm-ci.elastic.co/job/apm-integration-tests-selector-mbp/job/PR-763/26/)

### CODE REVIEW

File: .ci/scripts/opbeans-rum.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+## Test the service is up and running
+http_code=$(curl -s -o /dev/null -w "%{http_code}" http://0.0.0.0:3000/dashboard)
+if [ "${http_code}" == "200" ] ; then
+  docker logs opbeans-frontend | grep '/dashboard'
+  docker stop opbeans-frontend || true
+else
+  echo 'ERROR: service is not listening.'
+  exit 1
+fi
```
Comment (v1v) [inline .ci/scripts/opbeans-rum.sh]: A basic test to validate if the opbeans-frontend docker image does work as expected. Anything related to enhancing the test coverage will be done in the upcoming follow ups.

File: docker/opbeans/go/Dockerfile [hunk 2]
```diff
@@ -15,10 +15,11 @@ RUN rm -fr vendor/go.elastic.co/apm
 
 ARG GO_AGENT_REPO=elastic/apm-agent-go
 ARG GO_AGENT_BRANCH=master
-# adding this file will invalidate the layer cache when the branch head changes
-ADD https://api.github.com/repos/${GO_AGENT_REPO}/git/refs/heads/${GO_AGENT_BRANCH} version.json
+
 RUN git clone https://github.com/${GO_AGENT_REPO}.git /src/go.elastic.co/apm
-RUN (cd /src/go.elastic.co/apm && git checkout ${GO_AGENT_BRANCH})
+RUN (cd /src/go.elastic.co/apm \
```
Comment (v1v) [inline docker/opbeans/go/Dockerfile]: Support sha checkouts

File: docker/opbeans/dotnet/run.sh [hunk 3]
```diff
@@ -13,8 +13,11 @@ CSPROJ_VERSION="/src/dotnet-agent/src/Elastic.Apm.NetCoreAll/${PACKAGE}.csproj"
 BUILD_PROPS="/src/dotnet-agent/src/Directory.Build.props"
 
 if [ -z "${DOTNET_AGENT_VERSION}" ] ; then
-  git clone https://github.com/"${DOTNET_AGENT_REPO}".git /src/dotnet-agent -b "${DOTNET_AGENT_BRANCH}"
+  git clone https://github.com/"${DOTNET_AGENT_REPO}".git /src/dotnet-agent
   cd /src/dotnet-agent || exit
+  git fetch -q origin '+refs/pull/*:refs/remotes/origin/pr/*'
```
Comment (v1v) [inline docker/opbeans/dotnet/run.sh]: Support sha checkouts

File: .ci/integrationTestSelector.groovy [hunk 4]
```diff
@@ -155,19 +207,18 @@ pipeline {
   }
 }
 
-def wrappingup(){
+def wrappingup(Map params = [:]){
+  def isJunit = params.containsKey('isJunit') ? params.get('isJunit') : true
```
Comment (v1v) [inline .ci/integrationTestSelector.groovy]: There are not tests for the opbeans yet, so this is the way we can reuse the same method without the junit step.

File: .ci/integrationTestSelector.groovy [hunk 5]
```diff
@@ -51,29 +52,80 @@ pipeline {
...
+          agent { label 'linux && immutable' }
+          options { skipDefaultCheckout() }
+          when {
+            expression { return (params.AGENT_INTEGRATION_TEST == 'RUM') }
+            beforeAgent true
+          }
+          steps {
+            deleteDir()
+            unstash "source"
+            dir('opbeans-frontend') {
```
Comment (v1v) [inline .ci/integrationTestSelector.groovy]: Opbeans-fronted is consumed here, in other words, there is no an opbeans-frontend image in the apm-its.

Before testing the RUM  PR, it's required to build the docker image from the source code and run the bump version script to change the agent version accordingly.

File: .ci/integrationTestSelector.groovy [hunk 6]
```diff
@@ -51,29 +52,80 @@ pipeline {
...
     */
-    stage("Integration Tests"){
+    stage('Integration Tests'){
+      options { skipDefaultCheckout() }
       environment {
         TMPDIR = "${WORKSPACE}"
         ENABLE_ES_DUMP = "true"
         PATH = "${WORKSPACE}/${BASE_DIR}/.ci/scripts:${env.PATH}"
         APP = agentMapping.app(params.AGENT_INTEGRATION_TEST)
+        OPBEANS_APP = agentMapping.opbeansApp(params.AGENT_INTEGRATION_TEST)
```
Comment (v1v) [inline .ci/integrationTestSelector.groovy]: There is a match between agents and opbeans repos, but the `Node.js` opbeans app is called `node` instead.

Comment (mdelapenya) [reply .ci/integrationTestSelector.groovy]: https://github.com/elastic/apm-pipeline-library/issues/398 could help?

Comment (v1v) [reply .ci/integrationTestSelector.groovy]: It's now solved with the maps, but wanted to clarify one of the exceptions for avoid using the agent name

**Pull request merged**